### ISSUE TITLE
sudo hangs for 2min with no internet connection

### ISSUE THREAD

**Issue by aidanhs** [Microsoft/WSL#884]
Windows version 1607 build 14393.51. I'm a little surprised I couldn't find a duplicate issue, apologies if I missed it.

When running `sudo anycommand` with no internet connection, it will hang for a while and then eventually say it couldn't resolve the name of my machine (A-3).

```
root@A-3:~# time sudo ls
sudo: unable to resolve host A-3

real    2m4.961s
user    0m0.000s
sys     0m0.047s
```

strace reveals that it's attempting to hit the dns servers in resolv.conf

```
root@A-3:~# cat /etc/resolv.conf 
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 192.168.1.254
nameserver 192.11.128.24
nameserver fec0:0:0:ffff::1
search home
root@A-3:~# strace -f sudo ls
execve("/usr/bin/sudo", ["sudo", "ls"], [/* 16 vars */]) = 0
[...]
socket(PF_INET, SOCK_DGRAM|SOCK_NONBLOCK, IPPROTO_IP) = 6
connect(6, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.1.254")}, 16) = 0
gettimeofday({1471084539, 951278}, NULL) = 0
poll([{fd=6, events=POLLOUT}], 1, 0)    = 1 ([{fd=6, revents=POLLOUT}])
sendmmsg(6, {{{msg_name(0)=NULL, msg_iov(1)=[{"A\345\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\1\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}, {{msg_name(0)=NULL, msg_iov(1)=[{"\17\t\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\34\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=6, events=POLLIN}], 1, 5000)  = 0 (Timeout)
socket(PF_INET, SOCK_DGRAM|SOCK_NONBLOCK, IPPROTO_IP) = 7
connect(7, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.11.128.24")}, 16) = 0
gettimeofday({1471084550, 486769}, NULL) = 0
poll([{fd=7, events=POLLOUT}], 1, 0)    = 1 ([{fd=7, revents=POLLOUT}])
sendmmsg(7, {{{msg_name(0)=NULL, msg_iov(1)=[{"A\345\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\1\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}, {{msg_name(0)=NULL, msg_iov(1)=[{"\17\t\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\34\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=7, events=POLLIN}], 1, 3000)  = 0 (Timeout)
socket(PF_INET6, SOCK_DGRAM|SOCK_NONBLOCK, IPPROTO_IP) = 8
connect(8, {sa_family=AF_INET6, sin6_port=htons(53), inet_pton(AF_INET6, "fec0:0:0:ffff::1", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, 28) = 0
gettimeofday({1471084559, 152047}, NULL) = 0
poll([{fd=8, events=POLLOUT}], 1, 0)    = 1 ([{fd=8, revents=POLLOUT}])
sendmmsg(8, {{{msg_name(0)=NULL, msg_iov(1)=[{"A\345\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\1\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}, {{msg_name(0)=NULL, msg_iov(1)=[{"\17\t\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\34\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=8, events=POLLIN}], 1, 6000)  = 0 (Timeout)
gettimeofday({1471084577, 144768}, NULL) = 0
poll([{fd=6, events=POLLOUT}], 1, 0)    = 1 ([{fd=6, revents=POLLOUT}])
sendmmsg(6, {{{msg_name(0)=NULL, msg_iov(1)=[{"A\345\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\1\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}, {{msg_name(0)=NULL, msg_iov(1)=[{"\17\t\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\34\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=6, events=POLLIN}], 1, 5000)  = 0 (Timeout)
gettimeofday({1471084588, 149977}, NULL) = 0
poll([{fd=7, events=POLLOUT}], 1, 0)    = 1 ([{fd=7, revents=POLLOUT}])
sendmmsg(7, {{{msg_name(0)=NULL, msg_iov(1)=[{"A\345\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\1\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}, {{msg_name(0)=NULL, msg_iov(1)=[{"\17\t\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\34\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=7, events=POLLIN}], 1, 3000)  = 0 (Timeout)
gettimeofday({1471084594, 145744}, NULL) = 0
poll([{fd=8, events=POLLOUT}], 1, 0)    = 1 ([{fd=8, revents=POLLOUT}])
sendmmsg(8, {{{msg_name(0)=NULL, msg_iov(1)=[{"A\345\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\1\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}, {{msg_name(0)=NULL, msg_iov(1)=[{"\17\t\1\0\0\1\0\0\0\0\0\0\3A-3\4home\0\0\34\0\1", 26}], msg_controllen=0, msg_flags=0}, 26}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=8, events=POLLIN}], 1, 6000)  = 0 (Timeout)
close(6)                                = 0
close(7)                                = 0
close(8)                                = 0
socket(PF_INET, SOCK_DGRAM|SOCK_NONBLOCK, IPPROTO_IP) = 6
connect(6, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.168.1.254")}, 16) = 0
gettimeofday({1471084612, 157587}, NULL) = 0
poll([{fd=6, events=POLLOUT}], 1, 0)    = 1 ([{fd=6, revents=POLLOUT}])
sendmmsg(6, {{{msg_name(0)=NULL, msg_iov(1)=[{"\220?\1\0\0\1\0\0\0\0\0\0\3A-3\0\0\1\0\1", 21}], msg_controllen=0, msg_flags=0}, 21}, {{msg_name(0)=NULL, msg_iov(1)=[{"\346\205\1\0\0\1\0\0\0\0\0\0\3A-3\0\0\34\0\1", 21}], msg_controllen=0, msg_flags=0}, 21}}, 2, MSG_NOSIGNAL) = 2
poll([{fd=6, events=POLLIN}], 1, 5000)  = 0 (Timeout)
socket(PF_INET, SOCK_DGRAM|SOCK_NONBLOCK, IPPROTO_IP) = 7
connect(7, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("192.11.128.24")}, 16) = 0
gettimeofday({1471084620, 144464}, NULL) = 0
poll([{fd=7, events=POLLOUT}], 1, 0)    = 1 ([{fd=7, revents=POLLOUT}])
sendmmsg(7, {{{msg_name(0)=NULL, msg_iov(1)=[{"\220?\1\0\0\1\0\0\0\0\0\0\3A-3\0\0\1\0\1", 21}], msg_controllen=0, msg_flags=0}, 21}, {{msg_name(0)=NULL, msg_iov(1)=[{"\346\205\1\0\0\1\0\0\0\0\0\0\3A-3\0\0\34\0\1", 21}], msg_controllen=0, msg_flags=0}, 21}}, 2, MSG_NOSIGNAL) = 2
```

And presumably continues to do so until it eventually gives up.

Here's `/etc/hosts`:

```
root@A-3:~# cat /etc/hosts
127.0.0.1 localhost

# The following lines are desirable for IPv6 capable hosts
::1 ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
ff02::3 ip6-allhosts
```

And for comparison, an actual Ubuntu 14.04 on the same computer:

```
root@a-3-linux:~# cat /etc/hosts
127.0.0.1 localhost
127.0.1.1 a-3-linux

# The following lines are desirable for IPv6 capable hosts
::1 ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

So I guess the solution is to just inject the current hostname into /etc/hosts

**Comment by fpqc** [comment]
Delete /etc/hosts and restart bash (all windows). It will regenerate a fixed /etc/hosts.  This is a known bug.

**Comment by rodrymbo** [comment]
You could also just edit /etc/hosts to make sure `A-3` is given a localhost IP address there. That should happen automatically if you do as @fpqc suggests. At least you should check `/etc/hosts` after deleting it,  stopping all bash processes, and restarting bash, to make sure `A-3` is there.

If sudo isn't working, you won't be able to delete `/etc/hosts` without doing `lxrun /setdefaultuser root` (from CMD). You can set it back to your regular user when you're finished.

The other thing to note is that when sudo did that on other machines (reporting that it couldn't resolve the local host's name), it usually went ahead and did the task anyway.  Resolving localhost's hostname is easy, so by all means fix that (by editing `/etc/hosts`, or deleting it and letting WSL regenerate it), but there's a chance it is not the real problem.

**Comment by aidanhs** [comment]
@fpqc oh great, do you have a link somewhere? Thanks for the workaround, though I'm unable to test it since I can't get into the environment right now because of another issue :(

@rodrymbo I think I've given a wrong impression - the task does complete successfully, it just takes 2 mins to get there.

**Comment by GioBonvi** [comment]
This happened to me many times.
The issue seems to be the one you higlighted: when issuing a sudo command without an Internet connection (for example if WiFi is off) sudo tries to connect to the DNS servers, but somehow the lack of an Internet connection makes the process hang for 2 minutes more or less. After the two minutes it gives up connecting to the DNS and executes the command, although complaining that it was "unable to resolve the hostname".

The solution I used is editing the /etc/hosts file:

I changed the line reading
`127.0.0.1 localhost`
to
`127.0.0.1 localhost MYHOSTNAME`
Where MYHOSTNAME is what it shows after the username in the shell (es: `user@HOSTNAME:/home/user $`)

This solves the problem immediately.

The /etc/hosts file should be corrected to be installed directly like this, without the need to manually editing it or deleting it and restarting as @fpqc correctly suggested.
